enhancements/accelerators/out-of-tree-drivers.md

```diff
@@ -0,0 +1,416 @@
+ 
+#### Day-1 DriverContainer Management  [OpenShift | BareMetal]
+ 
+TODO: LONGER TERM
+ 
+#### Day-2 DriverContainer Management OpenShift
+ 
+More and more customers/partners want to enable hardware and or software on OpenShift that need kernel drivers, which are currently not (and may never be) upstream   and out-of-tree. 
+We want to provide a unified way to support out-of-tree kernel drivers (multiple) on OpenShift for day-1 (long term) and day-2 (short term). It has to work for classical (RHEL7, 8, Fedora) and container based operating systems (RHCOS, FCOS) in the same way. 
+Many customers/partners are using dkms/akmods as the solution to build, rebuild modules on kernel changes. Adopting dkms is not a solution; we need to  create an own acceptable solution for RHEL and Fedora. 
+ 
+#### Fill the gap of providing drivers that are not yet upstream or never be
+
+For the DriverContainer we need to cover several stages of driver packaging. 
+
+* **github repository or archive** The driver is only available as  source code and is not packaged in any way, this is where KVC can help
+* **kmod-{vendor}-src.rpm** The next step is a source RPM package that can be recompiled with rpmbuild. This is also the base for akmods, dkms
+* **precompiled-{vendor}-{kernelversion}.rpm** Precompiled RPMS this is the wish thinking to have in the future. DriverContainers could be easily built just by using RPMs. 
+
+Some drivers will *never* be upstreamed and can be in any of the states described above.  The proposed solutions needs to handle drivers in any state and build by "any" tool. 
+
+The compilation of kernel modules was always anticipated to be the fallback solution when dealing with kernel modules. Some kernel modules will always be out-of-tree and are not going to be included in the near future. Some kernel modules are out-of-tree but we are working with the vendors on upstreaming them  to the mainline kernel. 
+This work is being done by a virtual-team consisting of an  OpenShift-accelerators  and RHEL-accelerator team responsible for enabling hardware and software on RHEL and OpenShift. 
+
+
+### Goals
+
+
+* A unified way to deploy out-of-tree drivers on OpenShift 4.x on all supported Red Hat Operating Systems
+* The solution should avoid rebuilds on every node and allow for distribution of drivers on a cluster using the cluster registry
+* A solution for day-1 and day-2 kernel modules
+* Support upgrades of OpenShift Support of multiple kernel module providers
+* Hierarchical initialization of kernel modules (modeling dependencies)
+*  Handle dependencies between kernel modules (depmod) in tree and out of tree
+* Should support disconnected and proxy environments
+* Support building a DriverContainer on any host or kernel version (-k flags) e.g CI
+* Support heterogeneous cluster: 
+  * OpenShift with RHEL7, 8 and RHCOS 
+  * OKD is only supporting FCOS nodes for now. 
+* Extended Update Support
+
```

Comment:
We will have this functionality, it is possible now, e.g. we can have side-car containers for NFD to advertise more detailed information or use NFD to expose hardware as a extended-resource. We have created a new NFD release  (v0.6) with extended resource support and a general matching mechanism with rules.  See https://github.com/kubernetes-sigs/node-feature-discovery/issues/300 for all changes.

Reply:
ok, so we will have the choice of:
1) add this functionality in new operator by running a NFD side-car container
2) use NFD Operator directly to expose extended-resource of hardware enabled by third-party drivers.
but whether 1) or 2) is not decided.

Reply:
It will be 2. but the new operator will deploy the DriverContainer and optionally the side-car container for NFD. It is a two-phase deployment. 

1. DriverContainer (enable hardware, exposing device files, /sysfs, ..., build artifacts)
2. SIdeCar -- vendor feature discovery (exposing specific characteristics via NFD with labels) 

In the second stage (2) we have the drivers loaded and aux tools running,  so we can extract sophisticated information about the hardware.